Mage/src/main/java/mage/abilities/BatchTriggeredAbility.java

```diff
@@ -0,0 +1,29 @@
+
+package mage.abilities;
+
+import mage.game.Game;
+import mage.game.events.GameEvent;
+
+import java.util.stream.Stream;
+
+/**
+ * Batch triggers (e.g. 'When... one or more ..., ')
+ * are triggers that require a little more details.
+ *
+ * @author Susucr
+ */
+public interface BatchTriggeredAbility<T extends GameEvent> extends TriggeredAbility {
+
+    /**
+     * filter a batch event into all it's sub events that are relevant.
+     * <p>
+     * Properly filtering is required for further analysis of trigger + event,
+     * for instance for complex NUMBER_OF_TRIGGERS triggers.
+     * e.g. Umezawa's Jitte + Felix Five-Boots.
+     * <p>
+     * Note: if calling this, make sure the event does pass TriggeredAbility::checkEventType
+     */
+    Stream<T> filterBatchEvent(GameEvent event, Game game);
```

Comment:
I would prefer if we could split the stream logic and the filter logic into separate methods, something like:

```java

boolean checkEvent(T event, Game game);

default Stream<T> filterBatchEvent(BatchEvent<T> event, Game game) {
    return event.getEvents()
                .stream()
                .filter(e -> checkEvent(e, game))
}

```

with the checkTrigger methods starting with

`filterBatchEvent((TypeOfBatchEvent) event, game)`

but I'm not sure, will the generic types work for this sort of usage?

Reply:
Uhm I'll try something like that. May need another method to do the event casting a safe way, as the NUMBER_OF_TRIGGERS will not have direct access to the batch event type.

Reply:
Yes, must use overrideable method like `filterEvent` from xenohedron above -- so all implementation will override only simple boolean method and no needs in streams at all. Also I recommends to return List<GameEvent> in find method, not stream<T> -- it's a standard code style for xmage.